### PR TITLE
[Elastic Agent] Add doc on how Fleet Server is bootstrapped.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Adds developer documentation on how the Elastic Agent bootstraps the Fleet Server under the Elastic Agent.

## Why is it important?

To help other engineers on the team to understand the process.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- [x] I have made corresponding changes to the documentation
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
+[[fleet-server-bootstrap]]
+== Fleet Server Bootstrap
+
+Elastic Agent with Fleet Server has a bootstrap process that it uses to get
+Fleet Server up and running under Elastic Agent.
+
+Upon `install` or `enroll` with Elastic Agent will bootstrap a Fleet Server
+when the `--fleet-server-es` command-line option is provided. In this mode
+Elastic Agent will only communicate with its local Fleet Server and expose
+the Fleet Server over the `:8220` port.
```
Comment (jlind23) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:10]: Based on this: https://github.com/elastic/fleet-server/pull/967/files
Both 8220 and 8221 will be used. 8221 for exclusive communication with the local Elastic Agent running.

Comment (blakerouse) [reply x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc]: I updated to documentation to use :8221 and added description on the difference between :8220 and :8221.

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 2]
```diff
@@ -0,0 +1,90 @@
+[[fleet-server-bootstrap]]
+== Fleet Server Bootstrap
+
+Elastic Agent with Fleet Server has a bootstrap process that it uses to get
+Fleet Server up and running under Elastic Agent.
+
+Upon `install` or `enroll` with Elastic Agent will bootstrap a Fleet Server
+when the `--fleet-server-es` command-line option is provided. In this mode
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:8]:
```suggestion
Elastic Agent will bootstrap a Fleet Server when the `--fleet-server-es`
command-line option is provided to an `install` or `enroll` command
```

i think it's a bit easier to read

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+Elastic Agent will only communicate with its local Fleet Server and expose
+the Fleet Server over the `:8221` port.
+
+`:8221` is only bound to the localhost of the machine/container and cannot
+be accessed remotely. The `:8220` port is also bound but is bound on `0.0.0.0`
+to allow remote connections from external Elastic Agents that wish to enroll
+and communication. The `:8221` port is reserved for communication between the
+Elastic Agent and the Fleet Server on the host. This ensures that the local
+Elastic Agent has priority in check-ins with the Fleet Server.
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:17]:
```suggestion
The `:8221` port is reserved for communication between the Elastic Agent and the
Fleet Server on the host. It is bound to the localhost of the machine/container
and cannot be accessed remotely. This ensures that the local Elastic Agent has
priority in check-ins with the Fleet Server. The `:8220` port is bound on
`0.0.0.0` to allow remote connections from external Elastic Agents that wish to
enroll and communication.
```
re ordered to make it clear what the differences between ports are

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+Using the `--insecure` and `--fleet-server-insecure-http` will bootstrap the Fleet Server
+without any certificates, it will be bound to `localhost:8220` and Elastic Agent will
+communicate in clear-text.
+
+==== Custom Certificates (aka. Production)
+
+When deploying Elastic Agent in a production environment using enterprise generated
+certificates will ensure that Elastic Agent running locally and remote Elastic Agent
+will be able to connect over a verified TLS based connection. Passing in the certificates
+use `--fleet-server-cert`, `--fleet-server-cert-ca`, and `--certificate-authorities`.
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:46]:
```suggestion
will be able to connect over a verified TLS based connection. Certificates are specified
with `--fleet-server-cert`, `--fleet-server-cert-ca`, and `--certificate-authorities`.
```

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+[float]
+[[fleet-server-bootstrap-process]]
+=== How Does It Bootstrap
+
+Bootstrapping happens inside of the `enroll` command. Even when the `install` command
+or the `container` command (used by Docker container) is used they both still call
+the `enroll` command to perform the bootstrapping process.
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:54]:
```suggestion
Bootstrapping is ran during the `enroll` command. The `install` command
or `container` command (used by Docker container) will call the `enroll`
command to perform the bootstrapping process.

**Review by michel-laterman — COMMENTED** [review]
reworded some parts to try to make it a bit clearer

**Comment by blakerouse** [comment]
@michel-laterman Updated with your suggestions. Looking for you +1.

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+`0.0.0.0` to allow remote connections from external Elastic Agents that wish to
+enroll and communicate.
+
+[float]
+[[fleet-server-operating-modes]]
+=== Operating Modes
+
+Elastic Agent can bootstrap the Fleet Server into three different modes. In either
+mode the Fleet Server operates the same, but how it exposes itself over the `:8220`
+port is different.
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:25]:
```suggestion
Elastic Agent can bootstrap the Fleet Server into three different modes. The mode
determines how Fleet Server exposes itself over the `:8220` port, but does not change
any other behaviour.
```

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+==== Self-signed Certificate
+
+With the standard `--fleet-server-es` and `--fleet-server-service-token` options the
+Elastic Agent will generate a self-generated CA and certificate for communication with
+the Fleet Server that it starts under the Elastic Agent. These certificates are generated
```
Comment (michel-laterman) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:31]:
```suggestion
Elastic Agent will generate a CA and certificate for communication with
the Fleet Server that it starts. These certificates are generated
```

**Review by michel-laterman — APPROVED** [review]
lgtm. just another couple of cleanup items

File: x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+==== HTTP Only
+
+Using the `--insecure` and `--fleet-server-insecure-http` will bootstrap the Fleet Server
+without any certificates, it will be bound to `localhost:8220` and Elastic Agent will
+communicate in clear-text.
+
+==== Custom Certificates (aka. Production)
+
+When deploying Elastic Agent in a production environment using enterprise generated
```
Comment (efd6) [inline x-pack/elastic-agent/docs/fleet-server-bootstrap.asciidoc:43]: s/enterprise generated/enterprise-generated/

**Pull request merged**